### DISCUSSION TITLE
Consolidate useMutation API to use the object pattern?

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [TanStack/query#3127] [Ideas]
> ⚠️ _This more of a subjective API / developer experience discussion. No bug or code related issues here._

While browsing the [`react-query` docs for `useMutation`](https://react-query.tanstack.com/reference/useMutation#_top) the example shown is the version of `useMutation` that takes the `mutationFn` as it's first argument.

There is also the version of `useMutation` that takes in an object (`UseMutationOptions`), but is not shown on the website docs (_it is with typescript types & editor autocomplete_)

I think the `useMutation` call using the `UseMutationOptions` object might have a few more benefits than the other available ways to call the `useMutation` hook:
- readable arguments
- order of passed in args doesn't matter
- easier to maintain docs and examples with this one 1 pattern

![CleanShot 2021-12-23 at 08 35 58@2x](https://user-images.githubusercontent.com/6743796/147268625-e760154c-e20e-4aba-84e6-a5d2b633f16b.png)

I imagine that the reason we have multiple ways to call `useMutation` is for backwards compatibility & historical purposes. In the future, do you think it is worth or even feasible consolidating the `useMutationAPI`? 
- Maybe I'm asking dumb questions, but simply trying to understand the the perspective of a library maintainer as well 🙏 

What do you think about me updating the docs examples to use the passing of the object approach?

**Comment by TkDodo** [comment]
It was actually on our roadmap for v4 to remove all overloads and only use a single object for useQuery and useMutation. However, we got stopped by a known typescript limitation where TS has troubles with type inference in certain situations, which would be worse. For reference: https://github.com/microsoft/TypeScript/issues/43371#issuecomment-984112995

**Reply by cliffordfajardo** [reply]
Ahhh - thanks for adding context!

**Comment by zbigg** [comment]
I was looking whether "object syntax" variant `useMutation` is an official or accidental (in typings, but not documented) API and found this thread and it looks like not documenting it is only _by omission_.

But original question was unanswered:

> What do you think about me updating the docs examples to use the passing of the object approach?

Are we right? Shall we update doc to state that `useMutation` also can also be used with "the object syntax"?

I'm happy to open PR, it should be trivial.

**Reply by TkDodo** [reply]
for queries, its documented here: https://react-query.tanstack.com/guides/query-functions#using-a-query-object-instead-of-parameters

we could add something similar to the mutations page: https://react-query.tanstack.com/guides/mutations
